rfp-proposals/rfp-proposal-storage-helper.md

```diff
@@ -0,0 +1,142 @@
+# RFP Proposal: `Storage Helper Libraries`
+
+**Name of Project: Storage Helper Libraries
+
+**Link to RFP:** [Storage Helper Libraries](https://github.com/filecoin-project/devgrants/blob/master/rfps/rfp-storage-helper-libraries.md)
+
+**RFP Category:** `devtools-libraries`
+
+**Proposer:** [@deaswang](https://github.com/deaswang)
+
+**Do you agree to open source all work you do on behalf of this RFP and dual-license under MIT and APACHE2 licenses?:** "Yes"
+
+# Project Description
+
+go-filecoin is to handle operations at the detailed, protocol level. Storage Helper Libraries separate from the go-filecoin node but generally reusable across a variety of storage use cases and implementations.
+
+Storage Helper Libraries should include:
+
+- **Data Structure**: file or folder information for DAG
+
+```
+MetaData {
+  path string // file path and name.
+  chunkNum int // the number of chunk.
+  size int //the size of file.
+  cids []string // all chunk cid order by seq.
+}
+```
+
+if store folder, it will be one list of MetaData.
+
+- **File ingestion and splitting**: prepares files for storage deals in dag-compatible way. recurse through folders. split files that are larger than selected filecoin sector size into chunks according to dag.
+
+```
+AddPiece(file): add file to the dag. return one MetaData object.
+AddPieces(folder, filter): add files by regular filtering to the dag, return list of MetaData object.
+```
+
+- **Get files**: get file from filecoin network.
+
+```
+GetFile(cid, filename): get file by metaData cid, and store as filename.
+GetFiles(cid, folder): get files by metaData cid, and store into the folder.
+```
+
+- **Storage deal agent**: primary job would be to monitor deals and renew as needed. could also select from miners with the best reputations, desired price range(s), desired geographical distribution.
+
+```
+ProposeDeal(miner, ask, datacid, price, duration, location): propose one deal.
+GetDeal(cid): get deal status.
+RenewDeal(cid, miner, ask, price, duration, location): renew deal.
+CancelDeal(cid): cancel deal.
+```
+
+- **Node operate**: operate with go-filecoin node.
+
+```
+InitNode(nodeAddr): connect go-filecoin node for other operate.
+BindWallet(address): bind with the wallet use for pay for deal.
+UnbindWallet(): unbind current wallet.
+```
+
+- **Sector**: sector operate for storage helper.
```

Comment:
Can you say a little more about how this helps accomplish the goal of the project?